File: Data Structures & Algorithms/max-area-of-island/submission-1.cpp
class Solution {
    int directions[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
private:
    int dfs(vector<vector<int>>& grid, int r, int c) {
        if (r < 0 || c < 0 || r >= grid.size() || c >= grid[0].size() || grid[r][c] == 0) {
            return 0;
        }
        grid[r][c] = 0;
        int size = 1;
        for (int i = 0; i < 4; i++) {
            size += dfs(grid, r + directions[i][0], c + directions[i][1]);
        }
        return size;
    }
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int row = grid.size(), col = grid[0].size(), biggest = 0;
        for (int r = 0; r < row; r++){
            for(int c = 0; c < col; c++){
                if (grid[r][c] == 1) {
                    biggest = max(biggest, dfs(grid, r, c));
                }
            }
        }
        return biggest;
    }
};

File: Data Structures & Algorithms/search-for-word/submission-1.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        directions = [[0, 1], [1, 0], [-1, 0], [0, -1]]
        visited = [[False] * len(board[0]) for _ in range(len(board))]
        if len(word) < 1:
            return False

        def dfs(i, j, idx):
            if idx >= len(word):
                return True
            for direction in directions:
                row = i + direction[0]
                col = j + direction[1]
                if row < 0 or len(board) <= row or col < 0 or len(board[0]) <= col or visited[row][col] == True :
                    continue
                if board[row][col] == word[idx]:
                    visited[row][col] = True
                    if dfs(row, col, idx + 1):
                        return True
                    visited[row][col] = False


        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0]:
                    visited[i][j] = True
                    if dfs(i, j, 1):
                        return True
                    visited[i][j] = False

        return False